internal/tool/tool.go

```diff
@@ -0,0 +1,183 @@
+package tool
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"os"
+	"os/exec"
+	"path"
+	"strings"
+
+	codacy "github.com/codacy/codacy-engine-golang-seed/v6"
+	"github.com/samber/lo"
+)
+
+const sourceConfigFileName = ".semgrep.yaml"
+
+// New creates a new instance of Codacy Semgrep.
+func New() codacySemgrep {
+	return codacySemgrep{}
+}
+
+// Codacy Semgrep tool implementation
+type codacySemgrep struct {
+}
+
+// https://github.com/uber-go/guide/blob/master/style.md#verify-interface-compliance
+var _ codacy.Tool = (*codacySemgrep)(nil)
+
+// Run runs the Semgrep implementation
+func (s codacySemgrep) Run(ctx context.Context, toolExecution codacy.ToolExecution) ([]codacy.Result, error) {
+	configFile, err := getConfigurationFile(*toolExecution.Patterns, toolExecution.SourceDir)
+	if err == nil {
+		defer os.Remove(configFile.Name())
+	}
+
+	// filesToAnalyse, err := getListOfFilesToAnalyse(tool.Files, sourceDir)
+	// if err != nil {
+	// 	return nil, errors.New("Error getting files to analyse: " + err.Error())
+	// }
+
+	semgrepCmd := semgrepCommand(configFile, *toolExecution.Files, toolExecution.SourceDir)
+
+	semgrepOutput, semgrepError, err := runCommand(semgrepCmd)
+	if err != nil {
+		return nil, errors.New("Error running semgrep: " + semgrepError)
+	}
+
+	result := parseOutput(toolExecution.ToolDefinition, semgrepOutput)
+	return result, nil
+}
+
+type SemgrepOutput struct {
+	Results []SemgrepResult
+}
+
+type SemgrepResult struct {
+	CheckID       string          `json:"check_id"`
+	Path          string          `json:"path"`
+	StartLocation SemgrepLocation `json:"start"`
+	EndLocation   SemgrepLocation `json:"end"`
+	Extra         SemgrepExtra    `json:"extra"`
+}
+
+type SemgrepLocation struct {
+	Line int `json:"line"`
+}
+
+type SemgrepExtra struct {
+	Message     string `json:"message"`
+	RenderedFix string `json:"rendered_fix,omitempty"`
+}
+
+func configurationFromSourceCode(sourceFolder string) (string, error) {
+	filename := path.Join(sourceFolder, sourceConfigFileName)
+	contentByte, err := os.ReadFile(filename)
+	return string(contentByte), err
+}
+
+func writeToTempFile(content string) (*os.File, error) {
+	tmpFile, err := os.CreateTemp(os.TempDir(), "semgrep-")
+	if err != nil {
+		return nil, err
+	}
+	if _, err = tmpFile.Write([]byte(content)); err != nil {
+		return nil, err
+	}
+	if err := tmpFile.Close(); err != nil {
+		return nil, err
+	}
+
+	return tmpFile, nil
+}
+
+func getConfigurationFile(patterns []codacy.Pattern, sourceFolder string) (*os.File, error) {
+	// if no patterns, try to use configuration from source code
+	// otherwise default configuration file
+	if len(patterns) == 0 {
+		sourceConfigFileContent, err := configurationFromSourceCode(sourceFolder)
+		if err == nil {
+			return writeToTempFile(sourceConfigFileContent)
+		}
+
+		return nil, err
+	}
+
+	// TODO: generate configuration file from patterns and auto config file
```

Comment:
here we need to generate a file with the rule definitions for the enabled patterns. we can use the 'auto' configuration file from the registry and extract the enabled rules to a temporary file